### PR TITLE
fix(ext/flash): graceful server startup/shutdown with unsettled promises in mind

### PR DESCRIPTION
**Pull request by magurotuna**
This PR resets the revert commit made by #16610, bringing back #16383 which attempts to fix the issue happening when we use the flash server with `--watch` option enabled. 
Also, some code changes are made to pass the regression test added in #16610. These changes can be seen here: https://github.com/denoland/deno/compare/0f31256..99f9670

Closes https://github.com/denoland/deno/issues/16699

### CODE REVIEW

File: ext/flash/lib.rs [hunk 1]
```diff
@@ -1139,16 +1178,57 @@ fn run_server(
...
+  // Now the flash thread is about to finish, but there may be some unsettled
+  // promises in the main thread that will use the socket. To make the socket
+  // alive longer enough, we move its ownership to the main thread.
+  for (tok, socket) in sockets {
+    if let Some(sender) = socket_senders.remove(&tok) {
+      // Do nothing if the receiver has already been dropped.
+      _ = sender.send(socket);
+    }
+  }
+
```
Comment (magurotuna) [inline ext/flash/lib.rs:1201]: This is key; to prevent sockets from being released, transmit them to each ServerContext. This ensures that unsettled promises that outlive the flash thread can use sockets later.

File: ext/flash/socket.rs [hunk 2]
```diff
@@ -26,6 +29,7 @@ pub struct Stream {
   pub parse_done: ParseStatus,
   pub buffer: UnsafeCell<Vec<u8>>,
   pub read_lock: Arc<Mutex<()>>,
+  pub _pinned: PhantomPinned,
```
Comment (magurotuna) [inline ext/flash/socket.rs:32]: This is not directly related to this PR, but I feel this field is a must so we can make sure that `Stream` is not moved when it's `Pin`ned.

File: cli/tests/unit/flash_test.ts [hunk 3]
```diff
@@ -58,31 +58,38 @@ Deno.test(async function httpServerCanResolveHostnames() {
...
-    signal: ac.signal,
-    onListen: onListen(listeningPromise),
-    onError: createOnErrorCb(ac),
+    port: 0,
+    signal: ac1.signal,
+    onListen: (addr) => {
+      port = addr.port;
+      listeningPromise.resolve();
+    },
+    onError: createOnErrorCb(ac1),
```
Comment (magurotuna) [inline cli/tests/unit/flash_test.ts:74]: Using port number 4501 makes CI really flaky. On my local ubuntu machine I was not able to  see this flakiness though.

**Review by kt3k — APPROVED** [review]
Confirmed this change doesn't cause the issue https://github.com/denoland/deno_std/issues/2882 anymore

LGTM

**Comment by kt3k** [comment]
I found some test failures with net-related compat test cases.

If I run the following command in std, there seem many test failures:

```
path/to/target/debug/deno task node:test -- parallel/test-net
```

Look like this has a different problem? 🤔

**Comment by magurotuna** [comment]
ugh... Will dig into that. Minimum reproducible code would be appreciated.

**Comment by magurotuna** [comment]
What I noticed so far is the `close` op is called twice with the same resource ID. Still no idea the reason.

**Comment by magurotuna** [comment]
Super interesting, weird fact: if we build deno with `--release` option, the command `path/to/target/release/deno task node:test -- parallel/test-net` does SUCCEED. I confirmed this with several commits, including [d8aafcc](https://github.com/denoland/deno/pull/16616/commits/d8aafcc64e5670c06d526833d1f6e50e4f8fcc19) which is the latest commit in this branch. However, if we debug build, even 916598f8a77d1fb7b7feb577e600ead2bc45045f (which is v1.28.0) does not get these tests to pass. I have no idea at all why this weird thing is happening.

**Review by bartlomieju — APPROVED** [review]
LGTM, this fixes the error at hand in https://github.com/denoland/deno/issues/16699, so I'm gonna land it

**Comment by magurotuna** [comment]
Some test cases being flaky is a concern. I'll try and see if it's reproducible in my local Linux environment.

**Comment by magurotuna** [comment]
On my local ubuntu I cannot reproduce the flakiness on `httpServerRejectsOnAddrInUse` i.e. this test always passes as expected. I have no clue...

**Comment by magurotuna** [comment]
Really flaky 🤷

**Comment by bartlomieju** [comment]
Yeah, I'm not able to repeat this flakiness locally either :/

**Comment by magurotuna** [comment]
Hmm, from my quick observation, there seems to be three patterns of test failures.

1. `test release macos-12` timeout
2. `bench release ubuntu-20.04-xl` panic
3. `test release(debug) ubuntu-20.04-xl` fails at `httpServerRejectsOnAddrInUse` in `flash_test.ts`

Among these, only 3 is specific to this PR - other 2 failures happen on the main branch too.

**Comment by magurotuna** [comment]
[d7ca7eb](https://github.com/denoland/deno/pull/16616/commits/d7ca7eb7b5ea765ed7bb513c4a9b0c7f606eded2) fell into the case 1 (macos timeout)

**Comment by kt3k** [comment]
> Super interesting, weird fact: if we build deno with --release option, the command path/to/target/release/deno task node:test -- parallel/test-net does SUCCEED. I confirmed this with several commits, including [d8aafcc](https://github.com/denoland/deno/pull/16616/commits/d8aafcc64e5670c06d526833d1f6e50e4f8fcc19) which is the latest commit in this branch. However, if we debug build, even https://github.com/denoland/deno/commit/916598f8a77d1fb7b7feb577e600ead2bc45045f (which is v1.28.0) does not get these tests to pass. I have no idea at all why this weird thing is happening.

I also confirmed `deno task node:test -- parallel/test-net` doesn't have problem with `--release` build. I also found that debug build already have problem with these `net` related compat test cases in `main`. So those failures look irrelevant to this change (sorry for wasting your time!)

**Review by kt3k — APPROVED** [review]
LGTM

**Comment by magurotuna** [comment]
[67aac4a](https://github.com/denoland/deno/pull/16616/commits/67aac4a601abdfae5b9f0bafe173b8558842a557) fell into the case 3

**Comment by magurotuna** [comment]
@bartlomieju What do you think about the flakiness we are facing? I ran this failing test case locally line 10 times and saw no failure. Would it be okay to disable this test for the time being?

**Comment by bartlomieju** [comment]
> @bartlomieju What do you think about the flakiness we are facing? I ran this failing test case locally line 10 times and saw no failure. Would it be okay to disable this test for the time being?

Yes, I think that's reasonable

**Comment by magurotuna** [comment]
@bartlomieju PTAL before actually getting it landed just in case.

**Review by bartlomieju — APPROVED** [review]
LGTM
